packages/xstate-react/src/useActor.ts

```diff
@@ -1,24 +1,49 @@
 import { useState, useRef } from 'react';
 import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
-import { Sender, ActorRefLike } from './types';
-import { EventObject, Actor } from 'xstate';
+import { Sender } from './types';
+import { ActorRef } from 'xstate';
 import useConstant from './useConstant';
 
-export function useActor<TEvent extends EventObject, TEmitted = any>(
-  actorRef: ActorRefLike<TEvent, TEmitted> | Actor,
-  getSnapshot: (actor: typeof actorRef) => TEmitted = (a) =>
-    'state' in a ? a.state : (undefined as any)
-): [TEmitted, Sender<TEvent>] {
+function isActorWithState<T extends ActorRef<any>>(
+  actorRef: T
+): actorRef is T & { state: any } {
+  return 'state' in actorRef;
+}
+
+function isDeferredActor<T extends ActorRef<any>>(
+  actorRef: T
+): actorRef is T & { deferred: boolean } {
+  return 'deferred' in actorRef;
+}
+
+type EventOfActorRef<
+  TActor extends ActorRef<any, any>
+> = TActor extends ActorRef<infer TEvent, any> ? TEvent : never;
+type EmittedOfActorRef<
+  TActor extends ActorRef<any, any>
+> = TActor extends ActorRef<any, infer TEmitted> ? TEmitted : never;
+
+const noop = () => {
+  /* ... */
+};
+
+export function useActor<
+  TActor extends ActorRef<any, any> = ActorRef<any, any>
+>(
```

Comment:
Maybe worth adding overloads to support `<TContext, TEvent>`? We could do this:
1. the first overload - the complex one with conditional types
2. the second one - `<TContext, TEvent, TActorWithoutDefault>`, so people could pass a custom `ActorRef` if needed
3. the third one - `<TContext, TEvent>`, what would probably be used most of the time when typed explicitly, has to come last to act as a fallback

Reply:
I don't think we can add overloads just for generics - not sure if that's possible in TS.

Reply:
A lot of TS "stdlib" is using this pattern - totally possible to provide overloads with totally different generic types, different amounts of them etc. An adjusted demo from the other post:
[TS playground](https://www.typescriptlang.org/play?noEmitHelpers=true#code/JYOwLgpgTgZghgYwgAgPICMDO0Bu0A8AKgHzIDeAUMsiBAB5gBcyAFDnADYCuEzhAlMgC8pHAHtgAEwDcVZNChiozFguZwQAT0Ejk4qbOoIxAWwAOHCJBU7REmRQC+FCqEixEKAMpcsCKMBmYMBiIORyXCCYvpj+wOgQLPzM+g7OruDQ8EjIPn4B6HDolkSklNTR+fGJYli40MwY2FB4UKXJuTFxQSEghsiVsQWJctS0DCrs3LzIAsJ2UgA0o-JQilAA-CoKSupatnr2y9RGphZWEFusB6lyHXlDgcGhsuluWZ7IAIIIYEoAShAYEQAKJ4cDyBgQECSTDkZBgTRmGaYMABEAAc2QjkWsxBJmAYEgkmEyA0mmIcnokBhcIecUKxQgoIJRIgkjKcmwMO24KYeL5N3srxciORyDB0LAGAAVhBfqSyGKUWjQBj0pJ5Rw4FAUDBIr9esguNgfn82nJCGalJCabDvr8AUD8OTceTSEIHebAcDXWStJTiCw5IhvUC+NaoMdkBirF4QHAzJgABZiMBXFih3azSMHK2OqC26H2yM+l1aXGgGDQPGs4mkDa1wnE5DMWitCgdADa+fNRdpXqdvoryCrNcI+Ob7IbTbZJLbEFauNUfIjBf7JYLZbHhYnfLdAeQjb3UtbNEX0CFUgAuhRNQhtbrkPqQIbQsbTQX8JbJRDqcW4TIBEkRVdEsRxH863ZCggxDLdw0HKAyxPcBcQnKCOWjWMwHjRMUzTDMs2URDkN-MA0MnOdiDzSjiU7Zge1o9llwvcA+DIq9JFvO8tR1FBjCiMAyQLZhS2dIDlWYAByGAxDEKTsVxEAuBMBIoEpCgBNRZAu1ROBIFxbkuNJE0IEjTMC34TTQm03SwH0iAACZDOLRzrxMz9zXwCSQOk2T5MUmgVLUoMiKsrShLshyAGYXJhaL3M9UzI3wMTgR85E-LkhScSC1ToGIULLKAA)

Reply:
Oh cool! Didn't know you can do that 👍

Reply:
Upon trying this, adding the overload doesn't work well with inference.

We want `useActor(someActor)` to properly infer to `useActor<typeof someActor>(someActor)` but because of the overload, this will not work as expected.

Let's tackle in a separate PR if you think this is important; I am fine with having only `useActor<TActor>` supported for now; it doesn't make sense to separate `TEvent, TEmitted` when both of those are inherent parts of the same `TActor`.

Reply:
Could you take a look at this one: https://github.com/davidkpiano/xstate/pull/1668 ? Overloads are always matched from the top to the bottom and the matching stop on the first match so in a sense, it really shouldn't matter what overloads follow the signature with only `TActor` generic, and also the following overloads have no impact on the preceding ones.

Not a big deal but this would avoid a breaking change for the types - the second overload I've put there is matching what we have on master.